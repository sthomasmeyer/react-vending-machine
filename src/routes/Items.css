.Items {
  display: flex;
  margin-top: 2.5%;
}

.Items nav {
  border-right: solid 1px;
  padding: 1rem;
}

.Items-element {
  display: block;
  margin: 1rem 0;
}

.Items-element:hover {
  background-color: silver;
}
