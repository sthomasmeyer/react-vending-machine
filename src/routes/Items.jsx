import {
  Outlet,
  NavLink,
  useSearchParams,
  useLocation
} from 'react-router-dom';
import { getDrinks, getSnacks } from '../mockData';
import './Items.css';

const QueryNavLink = ({ to, ...props }) => {
  let location = useLocation();
  return <NavLink to={to + location.search} {...props} />;
};

const Items = ({ category }) => {
  let items;
  if (category === 'Drinks') {
    items = getDrinks();
  } else {
    items = getSnacks();
  }

  let [searchParams, setSearchParams] = useSearchParams();

  return (
    <div className='Items'>
      <nav>
        <input
          value={searchParams.get('filter') || ''}
          onChange={(evt) => {
            let filter = evt.target.value;
            filter ? setSearchParams({ filter }) : setSearchParams({});
          }}
        />
        {items
          .filter((item) => {
            let filter = searchParams.get('filter');
            if (!filter) return true;
            let name = item.name.toLowerCase();
            return name.startsWith(filter.toLowerCase());
          })
          .map((item) => (
            <QueryNavLink
              className='Items-element'
              to={`/${category}/${item.id}`}
              key={item.id}
            >
              {item.name}
            </QueryNavLink>
          ))}
      </nav>
      <Outlet />
    </div>
  );
};

export default Items;
