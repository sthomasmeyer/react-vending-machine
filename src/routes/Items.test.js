import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, fireEvent } from '@testing-library/react';
import Items from './Items';
import { MemoryRouter } from 'react-router-dom';

// Smoke test:
it('renders without crashing', () => {
  render(<Items category={'Snacks'} />, { wrapper: MemoryRouter });
});

it('matches snapshot', () => {
  const { asFragment } = render(<Items category={'Snacks'} />, {
    wrapper: MemoryRouter
  });
  expect(asFragment()).toMatchSnapshot();
});

it('generates expected HTML elements w/ given parameters', () => {
  render(<Items category={'Snacks'} />, { wrapper: MemoryRouter });

  expect(screen.getByText('Lori Harvey')).toBeInTheDocument();
  expect(screen.getByText('Snickers')).toBeInTheDocument();
  expect(screen.getByText(`Reese's Cups`)).toBeInTheDocument();
});

it('has a functional search mechanism', () => {
  render(<Items category={'Snacks'} />, { wrapper: MemoryRouter });

  expect(screen.getByText('Snickers')).toBeInTheDocument();
  expect(screen.getByText(`Reese's Cups`)).toBeInTheDocument();

  const searchInput = screen.getByDisplayValue('');
  fireEvent.change(searchInput, { target: { value: 'L' } });

  expect(screen.queryByText('Snickers')).not.toBeInTheDocument();
  expect(screen.queryByText('Snickers')).not.toBeInTheDocument();
  expect(screen.getByText('Lori Harvey')).toBeInTheDocument();
});

// it('displays info about a selected item', () => {
//   render(<Items category={'Snacks'} />, { wrapper: MemoryRouter });

//   expect(screen.queryByText(`Price: $2.22`)).not.toBeInTheDocument();

//   let targetElement = screen.getByText(`Reese's Cups`);
//   fireEvent.click(targetElement);

//   expect(screen.getByText(`Price: $2.22`)).toBeInTheDocument();
// });
