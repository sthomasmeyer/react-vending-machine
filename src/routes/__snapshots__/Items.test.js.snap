// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Items"
  >
    <nav>
      <input
        value=""
      />
      <a
        class="Items-element"
        href="/Snacks/04227dca-d83a-4356-96aa-de23a77398d4"
      >
        Reese's Cups
      </a>
      <a
        class="Items-element"
        href="/Snacks/6bf22ebf-8f52-414d-9e26-a86156dbcb9b"
      >
        Snickers
      </a>
      <a
        class="Items-element"
        href="/Snacks/df7d24d6-1625-4a73-9585-5dcc4c328645"
      >
        Lori Harvey
      </a>
    </nav>
  </div>
</DocumentFragment>
`;
